### PR TITLE
[Proposal] multi particle distribution

### PR DESCRIPTION
**Pull request by szabo137**
This is based on QEDjl-project/QEDprocesses.jl#23 and should be considered for merging after QEDjl-project/QEDprocesses.jl#23 is merged. 

# TODO

- [x] think about better implementation of type stable `rand` function. The current test implementation uses a hard return type, which is kinda ugly.

### CODE REVIEW

File: test/test_implementation/multi_particle_dist.jl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    return d.parts
+end
+
+QEDevents._particle_directions(d::TestMultiParticleDist) = d.dirs
+
+function Distributions.rand(
+    rng::AbstractRNG, d::TestMultiParticleDist{DT,PT,RT}
+)::RT where {DT,PT,RT}
+    return _groundtruth_multi_rand(rng, d)
+end
```
Comment (szabo137) [inline test/test_implementation/multi_particle_dist.jl]: Currently, I wouldn't say I like this implementation with the hard return type, but it is necessary for type stability. 

One idea to improve this would involve changing `ParticleSampleable` and adding another interface function: 

```julia
abstract type ParticleSampleable{F<:QEDlikeVariate,RT} <: Sampleable{F,Continous}

function rand(rng::AbstractRNG,d::ParticleSampleable{...,RT})::RT where {RT}
    moms = rand_mom(rng,d)
    out = build_out(d,moms)
    return out
end
```
where `rand_mom` is a new interface function (replacing `rand`), `build_out` can be implemented generically using the type-parameter `RT`. This enforces the developer of concrete implementations to include the return type `RT` as a type parameter, which is then used for type stability. Building the return type in the constructor could be easier by providing suitable helper functions. 

A concrete implementation of a distribution would then look like
```julia
struct TestDist{RT} <: ParticleSampleable{...,RT}
    ...
    function TestDist(...)
        return_type = out_type_helper(...)
        new{return_type}(...)
    end
end
```
where `out_type_helper` could be implemented generically as well. It is essentially the current implementation of `Base.eltype`.

@AntonReinhard what do you think? Is it worth it to go in this direction?

File: src/interfaces/multi_particle_distribution.jl [hunk 2]
```diff
@@ -0,0 +1,119 @@
+
+"""
+
+   MultiParticleDistribution
+
+Base type for sample drawing from multiple particle distributions. The following interface functions
+should be implemented:
+
+* [`QEDevents._particles(d::MultiParticleDistribution)`](@ref): return
```
Comment (AntonReinhard) [inline src/interfaces/multi_particle_distribution.jl]: What is that `return` supposed to mean?

File: src/interfaces/single_particle_distribution.jl [hunk 3]
```diff
@@ -6,9 +6,9 @@
 Base type for sample drawing from single particle distributions. The following interface functions
 should be implemented:
 
-
-* [`QEDevents._particle(d::SingleParticleDistribution)`](@ref)
-* [`QEDevents._particle_direction(d::SingleParticleDistribution)`](@ref)
+* [`QEDevents._particle(d::SingleParticleDistribution)`](@ref): return associated particle
+* [`QEDevents._particle_direction(d::SingleParticleDistribution)`](@ref): return associated particle direction
+* `Distributions.rand(rng::AbstractRNG,d::SingleParticleDistribution)`: return random sample as a `ParticleStateful`
```
Comment (AntonReinhard) [inline src/interfaces/single_particle_distribution.jl]: Could you cherry-pick these changes into the previous merge request #23 ?

File: src/interfaces/single_particle_distribution.jl [hunk 4]
```diff
@@ -49,11 +49,11 @@ Interface function, which asserts that the given `input` is valid.
...
 ) where {D,P}
-    eltype(_particle_direction(d)) == D ||
-        throw(InvalidInputError("expected $(_particle_direction(d)) but got $D"))
+    typeof(_particle_direction(d)) == D ||
+        throw(InvalidInputError("expected $(typeof(_particle_direction(d))) but got $D"))
 
-    eltype(_particle(d)) == P ||
-        throw(InvalidInputError("expected $(_particle(d)) but got $P"))
+    typeof(_particle(d)) == P ||
+        throw(InvalidInputError("expected $(typeof(_particle(d))) but got $P"))
```
Comment (AntonReinhard) [inline src/interfaces/single_particle_distribution.jl]: Should also be in #23 if it's not too much work.

File: test/interfaces/multi_particle_distribution.jl [hunk 5]
```diff
@@ -0,0 +1,140 @@
+
+using QEDprocesses
+using QEDevents
+using QEDbase
+using Random: Random
+import Random: AbstractRNG, MersenneTwister, default_rng
+
+# only imported, because we want to test,
+# if QEDevents works without this (epsecially the Base.rand which is exported by
+# Distributions)
```
Comment (AntonReinhard) [inline test/interfaces/multi_particle_distribution.jl]:
```suggestion
# only imported because we want to test if QEDevents works without this 
# (especially the Base.rand, which is exported by Distributions)
```

File: test/interfaces/multi_particle_distribution.jl [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    test_dist = TestImpl.TestMultiParticleDist(test_directions, test_particles)
+
+    @testset "static properties" begin
+        @test @inferred QEDevents._particles(test_dist) == test_particles
+        @test @inferred QEDevents._particle_directions(test_dist) == test_directions
+        @test @inferred length(test_dist) == N
+        @test @inferred size(test_dist) == (N,)
+
+        # todo: consider to move _assemble_tuple_types to the test implementation
+        # (groundtruths must not rely on package internals)
```
Comment (AntonReinhard) [inline test/interfaces/multi_particle_distribution.jl:44]:
```suggestion
        # todo: consider to move _assemble_tuple_types to the test implementation
        # (groundtruths must not rely on package internals)
        # See https://github.com/QEDjl-project/QEDprocesses.jl/issues/75
```
I think https://github.com/QEDjl-project/QEDprocesses.jl/issues/75 would address this.

File: test/interfaces/multi_particle_distribution.jl [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        Random.seed!(RND_SEED)
+        psf_default = @inferred rand(test_dist)
+
+        @test psf_groundtruth == psf_rng
+        @test psf_rng == psf_default
+    end
+
+    @testset "multiple samples" begin
+        @testset "$dim" for dim in (1, 2, 3)
+            checked_lengths = (1, rand(RNG, 1:10))
```
Comment (AntonReinhard) [inline test/interfaces/multi_particle_distribution.jl]: This does not seem like the place to use RNG, especially since it's only called once. I think we should use some reasonable static length or iterate over a few different ones.

File: test/interfaces/single_particle_distribution.jl [hunk 8]
```diff
@@ -86,4 +88,29 @@ end
...
+            )
+            psf_wrong = ParticleStateful(
+                WrongDirection(), WrongParticle(), rand(RNG, SFourMomentum)
+            )
+
+            @test_throws InvalidInputError weight(test_dist, psf_wrong_particle)
+            @test_throws InvalidInputError weight(test_dist, psf_wrong_direction)
+            @test_throws InvalidInputError weight(test_dist, psf_wrong)
+        end
+    end
```
Comment (AntonReinhard) [inline test/interfaces/single_particle_distribution.jl]: As  above, should be cherry-picked into #23 if possible

File: test/test_implementation/multi_particle_dist.jl (hunk 1 above)
Comment (AntonReinhard) [reply test/test_implementation/multi_particle_dist.jl]: If I'm not mistaken, `build_out` would (more or less) be taken care of by introducing a `ParticleList` or similar as in https://github.com/QEDjl-project/QEDprocesses.jl/issues/75 .

However, I'm not entirely sure what prevents you from simply doing the `_assemble_tuple_types` in the `rand` call? Since it's only working on the types it should have zero overhead and yield a type-stable function all the same. You would also not need the extra constructor and the `RT` in the Distribution. With your existing wrapper of `Base.eltype` this should be trivial.

File: test/test_implementation/multi_particle_dist.jl [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+struct TestMultiParticleDist{DT<:Tuple,PT<:Tuple,RT} <: MultiParticleDistribution
+    dirs::DT
+    parts::PT
+    function TestMultiParticleDist(dirs::DT, parts::PT) where {DT,PT}
+        res_type = Tuple{QEDevents._assemble_tuple_types(parts, dirs, SFourMomentum)...}
+        return new{DT,PT,res_type}(dirs, parts)
+    end
+end
+
+#Base.length(d::TestMultiParticleDist) = length(d.parts)
```
Comment (AntonReinhard) [inline test/test_implementation/multi_particle_dist.jl]: Uncomment or delete this.

**Review by AntonReinhard — COMMENTED** [review]
I think the cherry-picks I mentioned in my comments would be resolved by moving PR #23 to commit f7c59902b2c2f4cbd7db1eb65b32c2d98e859d5f

So if I understand correctly, this PR adds an interface for sampling collections of some number and type of particles. I assume an interface for full `PhaseSpacePoint` sampling will follow as well? Then we'd have a relations between our phase space types and the distributions/samplers:
- `ParticleStateful` -> single particle distribution
- `ParticleList` -> multi particle distribution
- `PhaseSpacePoint` -> phase space point distribution (or something like that)

File: test/test_implementation/multi_particle_dist.jl (hunk 1 above)
Comment (szabo137) [reply test/test_implementation/multi_particle_dist.jl]: You are right, of course, if the constructor of a potential `ParticleList` can wrap a tuple or vector of `SFourMomenta` with the respective `AbstractParticleType` and `ParticleDirection` without much overhead. 

The problem with `_assemble_tuple_type` is that it implicitly builds the constructors of `ParticleStateful`, i.e. `ParticleStateful{Direction,Particle,MomType}`. But there is currently no constructor, which allows implicit wraps around a given momentum: something like `ParticleStateful{Dir,Part,MomType}(mom::MomType)`. I played around with it, but such constructors lead to weird implementations, where you need to call type parameters to construct the respective singleton object. 

In e3e02fd117b8d2977b93f5cd07c2636768a9f00b an explicit cast of a vector of four-momenta into a tuple of `ParticleStateful` is added. This makes the usage of return types and implicit casts obsolete and adds only a little overhead if the length of the tuples is not known at compile time. There are even better ways to implement this, but this would involve using generated functions to unroll the tuple construction, which is, IMHO, currently unnecessary.

File: test/interfaces/multi_particle_distribution.jl (hunk 7 above)
Comment (szabo137) [reply test/interfaces/multi_particle_distribution.jl]: I usually do this, if I want to test some random number *bigger* than one, where I don't want to rely on a specific number. 

BTW: there was a little bug in there, the drawn length could also be one. Fixed this in 31c9fa1 and added the prevalent case of length two as a case.

File: test/test_implementation/multi_particle_dist.jl (hunk 9 above)
Comment (szabo137) [reply test/test_implementation/multi_particle_dist.jl]: done in https://github.com/QEDjl-project/QEDevents.jl/commit/31c9fa1212696a29f21e4324cb269ac3c0d70fe0

File: src/interfaces/multi_particle_distribution.jl (hunk 2 above)
Comment (szabo137) [reply src/interfaces/multi_particle_distribution.jl]: nothing, removed it 🤷‍♂️

File: src/interfaces/single_particle_distribution.jl (hunk 3 above)
Comment (szabo137) [reply src/interfaces/single_particle_distribution.jl]: done and with the rebase it's gone from the changes.

File: src/interfaces/single_particle_distribution.jl (hunk 4 above)
Comment (szabo137) [reply src/interfaces/single_particle_distribution.jl]: done and with the rebase it's gone from the changes.

File: test/interfaces/single_particle_distribution.jl (hunk 8 above)
Comment (szabo137) [reply test/interfaces/single_particle_distribution.jl]: done and with the rebase it's gone from the changes.

File: test/interfaces/multi_particle_distribution.jl (hunk 7 above)
Comment (AntonReinhard) [reply test/interfaces/multi_particle_distribution.jl]: What I meant is that now it looks like it's testing numbers 1 to 10, but it actually tests 1, 2, and one unknown one of 3:10. So why not just test all of 3:10 if we expect it to work for all of those anyway? Or if that takes too much runtime, just the edge cases that are most likely to break or give useful information, e.g. 3 and 10.

Comment (szabo137) [reply test/interfaces/multi_particle_distribution.jl]: Understood, but there is no edge case except one or maybe two. This should just check if it works for something bigger than two. Just using one case, e.g. three, could be misleading for future selves. 

Edit: you are correct, checking all cases takes too long, and even the upper limit of ten is arbitrary in this case.

File: src/interfaces/multi_particle_distribution.jl [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    return Tuple(fill(UnknownDirection(), length(d)))
+end
+
+"""
+
+    randmom(rng::AbstractRNG,d::MultiParticleDistribution)
+
+Return an iteratable container (e.g. vector or tuple) of momenta according to the distribution `d`.
+"""
+function randmom end
```
Comment (AntonReinhard) [inline src/interfaces/multi_particle_distribution.jl]:
```suggestion
"""

    rand_mom(rng::AbstractRNG, d::MultiParticleDistribution)

Return an iterable container (e.g. vector or tuple) of momenta according to the distribution `d`.
"""
function rand_mom end
```

I know functions such as `is<thing>` are supposed to be without an underscore according to the Julia style guide, but does this apply here? I feel like `rand_mom(...)` would look nicer.

**Review by AntonReinhard — DISMISSED** [review]

File: src/interfaces/multi_particle_distribution.jl (hunk 10 above)
Comment (szabo137) [reply src/interfaces/multi_particle_distribution.jl]: I was following the convention of the Julia standard library: there they have `randn` and `randexp` without the underscore. 

Edit: I will open an issue, that `randmom` will used more generally, maybe directly on the level of `ParticleSampleable`. Then also `SingleParticleDistribution` and potentially the `ProcessDistribution` can benefit.

Comment (AntonReinhard) [reply src/interfaces/multi_particle_distribution.jl]: Okay

File: src/interfaces/multi_particle_distribution.jl [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+#default
+function _particle_direction(d::MultiParticleDistribution)
+    return Tuple(fill(UnknownDirection(), length(d)))
+end
+
+"""
+
+    randmom(rng::AbstractRNG,d::MultiParticleDistribution)
+
+Return an iteratable container (e.g. vector or tuple) of momenta according to the distribution `d`.
```
Comment (AntonReinhard) [inline src/interfaces/multi_particle_distribution.jl]:
```suggestion
Return an iterable container (e.g. vector or tuple) of momenta according to the distribution `d`.
```
(Sorry, just this last thing ^^)

File: src/interfaces/multi_particle_distribution.jl (hunk 10 above)
Comment (szabo137) [reply src/interfaces/multi_particle_distribution.jl]: Approved by authority 😅

File: src/interfaces/multi_particle_distribution.jl (hunk 11 above)
Comment (szabo137) [reply src/interfaces/multi_particle_distribution.jl]: Do not be sorry, it's my bad 😅

**Review by AntonReinhard — APPROVED** [review]
Ready to merge from my side

File: src/interfaces/multi_particle_distribution.jl (hunk 10 above)
Comment (AntonReinhard) [reply src/interfaces/multi_particle_distribution.jl]: I mean if it's part of the style guide then that's how it should be. I was just unaware of that :+1:
